embassy-stm32/build.rs

```diff
@@ -81,6 +81,16 @@ fn main() {
         singletons.push(c.name.to_string());
     }
 
+    // Extra analog switch pins available on most H7 chips
```

Comment:
this should check the pins actually exist for the currently-being-built chip

Reply:
That would be good. But how?
Enabling the split feature if there is no _C pin should fail asap, anyway. At the latest when modifying the syscfg in init().

Reply:
Letting a random register write deep in embassy fail to compile is not great, the compiler error ideally should tell you you've enabled an incompatible set of features.

Currently checking if a pin exists needs scanning all the gpio mappings to see if `PC2_C` or similar is mentioned somewhere. That signals it "exists".

Alternatively we could add a "pins: [...]" array to stm32-data that would list all pins that exist (and perhaps some properties of them, like package location?). This is something I've always wanted to do, because there are chips where only "part" of GPIO ports are implemented, this would allow preserving that. (Currently we assume if GPIOx exists then all Px[0..15] exist). It's a more involved refactor though, so perhaps not for now.

Reply:
Having package information in there would be ideal. As we know now, some packages even expose the PC2/3 Pins (actually only one, the huge BGA package). And for those, closing the analog switch and splitting the pins should probably not go hand in hand.

But yeah, that's going to be a bigger set of changes. Do you have a suggestion what am intermediate solution/compiler error should look like? (Or would the pin/signal scanning be sufficient already?)

Another quick solution that comes to mind:
We could catch 99% of the cases with something like "if (split feature enabled and not stm32h7 enabled) then throw compiler error".
Still, there are some H7s that have the PA0/1 with analog switch and afaik also some H7 that don't have any. Just the usual STM32 mess...

Reply:
> Or would the pin/signal scanning be sufficient already?

yes, that will work nicely. Just collect all seen pin names in a set, then error if `split-pc2` is enabled and the set doesn't contain `PC2_C`.

Reply:
Alright, I should have some time to do this tomorrow.

Reply:
Argh, while implementing it I realized that I never properly tested with the split activated. There need to be more changes to how the `pin_table` is generated.

Is this what it is supposed to look like? Then if so, I would put that in a loop.
```rust
                    if pin_name == "PC2" {
                        pins_table.push(vec![
                            "PC2_C".to_string(),
                            p.name.to_string(),
                            port_num.to_string(),
                            "2".to_string(),
                            format!("EXTI{}", 2),
                        ]);
                    }
```

I will push the changes with the "scan pins and put into a set" to check the validity of the features.
The code that iterates over Portletter + `for pin_num in 0u32..16 ` is another story. I can look into it in 2 weeks. For now I suggest leaving the PR open as WIP.